### PR TITLE
warn if we find multiple clippy configs

### LINKED ISSUE #8323: Clippy should probably say something if it sees both a `clippy.toml` and a `.clippy.toml`
### Description

I recently wasted a bit of time trying to figure out why my lint configurations in a `clippy.toml` file were not being honored by Clippy. It turned out that the problem was that the workspace already had a `.clippy.toml` file, which I hadn't noticed existed.

It appears that when clippy searches for config files, it just picks the first one it finds, and stops looking: https://github.com/rust-lang/rust-clippy/blob/8bb4690487a39f9faab77b308bd797a15c216703/clippy_lints/src/utils/conf.rs#L326-L335. So, it was picking the `.clippy.toml` file and ignoring the `clippy.toml` I had created.

Admittedly, this was a pretty silly mistake on my part! But, it would be nice if Clippy could notice when there are multiple config files in a workspace, and say "hey, looks like you have multiple config files in the same workspace, I'm only honoring this one". That would have saved me some confusion.

### Version

```text
rustc 1.56.1 (59eed8a2a 2021-11-01)
binary: rustc
commit-hash: 59eed8a2aac0230a8b53e89d4e99d55912ba6b35
commit-date: 2021-11-01
host: x86_64-unknown-linux-gnu
release: 1.56.1
LLVM version: 13.0.0
```

### Additional Labels

@rustbot label +C-enhancement

### PR DESCRIPTION
**Pull request by matthiaskrgr**
Fixes #8323

---

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog: warn if we find multiple clippy configs

### CODE REVIEW

**Comment by matthiaskrgr** [comment]
Hm, I'm actually wondering now if there are projects that have one clippy config in a root directory and another one in a subproject and whether we should warn in this case or not.. :thinking:

File: clippy_lints/src/utils/conf.rs [hunk 1]
```diff
@@ -322,18 +322,36 @@ pub fn lookup_conf_file() -> io::Result<Option<PathBuf>> {
...
                     Ok(md) if md.is_dir() => {},
-                    Ok(_) => return Ok(Some(config_file)),
+                    Ok(_) => {
+                        // warn if we happen to find two config files
+                        if let Some(ref found_config_) = found_config {
+                            eprintln!(
+                                "Warning: found two config files: {} and {}.\nUsing the first one!",
+                                found_config_.display(),
+                                config_file.display(),
+                            );
```
Comment (xFrednet) [inline clippy_lints/src/utils/conf.rs:342]: Could we maybe test this in our `ui-toml` test? :thinking:

And a small NIT, the names could be put into backticks. The second message also sounds a bit off to me. The messages "The first one will be used" or "Clippy will use the first one" sound a bit better to me :upside_down_face:.

**Comment by xFrednet** [comment]
> Hm, I'm actually wondering now if there are projects that have one clippy config in a root directory and another one in a subproject and whether we should warn in this case or not.. thinking

I was also wondering about then when I saw the code. I think those configs should actually be merged. Someone might want to deny functions, macros, and names on a workspace level and have some other configuration for a project. I feel like the configs from the workspace should propagate to projects unless they are overridden :thinking:

**Comment by xFrednet** [comment]
And it looks like high-five bot didn't assign anyone to this one. It looks interesting, I'll ~procrastinate~ assign my self :upside_down_face: 

r? @xFrednet

**Comment by camsteffen** [comment]
IMO it's unusual and sub-optimal to continue searching for a config after finding one. Instead we should document which filename takes precedence and have a way (verbose mode?) to print which config file is unused.

In general we can take inspiration from rustfmt ([source](https://github.com/rust-lang/rustfmt/blob/a21f1b6c2a5734f39a1efe3fa84d6475843d14fe/src/config/mod.rs#L355)).

**Comment by xFrednet** [comment]
I have no strong opinions on how workspace and project configuration files should be handled. I think there are compelling arguments for either side. One thing that speaks for only using one configuration file is that this is the current behavior (I think). In either case, we should definitely document how Clippy selects the configuration file. Thanks for pointing that out :upside_down_face:.

Also cc: @rust-lang/clippy maybe someone else has some thoughts on this :upside_down_face:

**Comment by Manishearth** [comment]
Ideally it would be nice to merge configs too. And we should probably support `.config/clippy.toml`.

But I'm fine with landing this as-is too

**Comment by matthiaskrgr** [comment]
I think merging configs and figuring out hierarchies probably is a whole other issue.

I think a separate `--debug-configs` or something like that just to print a second path into the terminal is a little overkill, but I would be find with perhaps making the warning a little less dramatic, like "Note: another config file at {} will be ignored"

**Comment by xFrednet** [comment]
> I think merging configs and figuring out hierarchies probably is a whole other issue.

Agreed, I didn't expect it to be done in this PR, it's just something that came up :upside_down_face:.

You could also change to note to just clearly state: "Clippy will be using {} as a config file, others will be ignored". I think it wouldn't hurt to specify it every time. Either message is fine, the current version also works IMO. :upside_down_face:

**Comment by flip1995** [comment]
> In general we can take inspiration from rustfmt ([source](https://github.com/rust-lang/rustfmt/blob/a21f1b6c2a5734f39a1efe3fa84d6475843d14fe/src/config/mod.rs#L355)).

I think it's a good idea to handle the Clippy config file in the same way as other config files are handled in the Rust ecosystem. I think a warning if a config file gets ignored is a nice improvement though.

**Comment by xFrednet** [comment]
Hey @matthiaskrgr, besides the discussion, have you seen my review comment at the start? Once that one is resolved, I think we can merge this PR :upside_down_face:

**Comment by matthiaskrgr** [comment]
I've tweaked the message a bit: `Using config file `{}`\nWarning: `{}` will be ignored.",`
and added some tests.

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
Looks good to me, thank you to the changes :upside_down_face: 

@bors r+

**Comment by xFrednet** [comment]
```diff
diff of stderr:

-Using config file `$SRC_DIR/tests/ui-cargo/multiple_config_files/warn/.clippy.toml`
-Warning: `$SRC_DIR/tests/ui-cargo/multiple_config_files/warn/clippy.toml` will be ignored.
+Using config file `/?/$SRC_DIR/tests/ui-cargo/multiple_config_files/warn/.clippy.toml`
+Warning: `/?/$SRC_DIR/tests/ui-cargo/multiple_config_files/warn/clippy.toml` will be ignored. 
```

It seems like windows is not quite happy with the `.stderr` file :thinking:

**Comment by matthiaskrgr** [comment]
I canonicalized the paths now, no idea if that helps but worth a shot. :shrug:

**Comment by xFrednet** [comment]
We'll give it a try :upside_down_face: 

@bors r+

**Comment by xFrednet** [comment]
Well, that doesn't seem to be working. We might have to restrict the tests to only one OS if the replacement doesn't work correctly :thinking:

**Comment by matthiaskrgr** [comment]
:/
I'm wondering why the other tests don't have this problem.
Is there some kind of trick?

**Comment by xFrednet** [comment]
I'm not sure. Taking a deep dive into compiletest-rs is on my todo list, but so are countless other things as well ^^. I can imagine that the replacement regex tries to only target paths in compiler message and might not be ideal for this case, but I can be wrong.

**Comment by matthiaskrgr** [comment]
I tried to fiddle a bit with relative paths but couldn't work out a solution (we'd have to un-relativize the path somehow..??) so I just ignored the test on windows now :I

File: tests/ui-cargo/multiple_config_files/warn/src/main.stderr [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// ignore-windows
```
Comment (xFrednet) [inline tests/ui-cargo/multiple_config_files/warn/src/main.stderr:1]: The CI complains about this comment. Doesn't this have to be added to the `.rs` test file? :upside_down_face:

Comment (matthiaskrgr) [reply tests/ui-cargo/multiple_config_files/warn/src/main.stderr:1]: Ah good call :D

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
Alright, still looks good to me, let's see what bors thinks

@bors r+

**Pull request merged**